src/decode/stream.rs

```diff
@@ -0,0 +1,497 @@
+use crate::decode::lzbuffer::{LZBuffer, LZCircularBuffer};
+use crate::decode::lzma::{
+    new_circular, new_circular_with_memlimit, DecoderState, LZMAParams, MAX_HEADER_LEN, START_BYTES,
+};
+use crate::decode::rangecoder::RangeDecoder;
+use crate::decompress::Options;
+use crate::error::Error;
+use std::io::{BufRead, Cursor, Read, Write};
+
+/// Maximum number of bytes to buffer while reading the header.
+const MAX_TMP_LEN: usize = MAX_HEADER_LEN + START_BYTES;
+
+/// Internal state of this streaming decoder. This is needed because we have to
+/// initialize the stream before processing any data.
+#[derive(Debug)]
+enum State<W>
+where
+    W: Write,
+{
+    /// Stream is initialized but header values have not yet been read.
+    Init(W),
+    /// Header values have been read and the stream is ready to process more data.
+    Run(RunState<W>),
+}
+
+/// Error type used to return the currently owned state along with the error on failure.
+struct StreamStateError<W>
+where
+    W: Write,
+{
+    /// State that was owned while the error occured
+    state: Option<State<W>>,
+    /// IO error that occured in that state
+    error: Error,
+}
+
+impl<W> StreamStateError<W>
+where
+    W: Write,
+{
+    /// Creates a new StreamStateError
+    fn new(state: Option<State<W>>, error: Error) -> Self {
+        Self { state, error }
+    }
+}
```

Comment:
Given that this is only used internally, I don't think this `new` function makes things clearer. In general `StreamStateError::new(foo, bar)` is less readable than `StreamStateError { state: foo, error: bar }`, because the function doesn't retain the field names.

Reply:
Generally I like adding `new` functions to initialize all structs. This makes it easy to change initialization in the future if the struct changes.

However, I also don't mind switching it to struct initialization.